File: backend/src/models/tontineModel.js
const pool = require('../config/database');

const create = async (tontineData) => {
    const {
        nom, description, montant_cotisation, frequence,
        nombre_membres_max, date_debut, code_invitation, created_by
    } = tontineData;

    const { rows } = await pool.query(
        `INSERT INTO tontines (
      nom, description, montant_cotisation, frequence,
      nombre_membres_max, date_debut, code_invitation, created_by
    ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
    RETURNING *`,
        [nom, description, montant_cotisation, frequence, nombre_membres_max, date_debut, code_invitation, created_by]
    );
    return rows[0];
};

const findAll = async (options = {}) => {
    let query = `
    SELECT t.* FROM tontines t
    JOIN memberships m ON t.id = m.tontine_id
    WHERE m.user_id = $1
  `;
    const params = [options.userId];
    let idx = 2;

    if (options.statut) {
        query += ` AND t.statut = $${idx}`;
        params.push(options.statut);
        idx++;
    }

    query += ' ORDER BY t.date_creation DESC';
    const { rows } = await pool.query(query, params);
    return rows;
};

const findById = async (id, userId) => {
    // Get basic tontine info
    const { rows } = await pool.query(
        `SELECT t.*, m.role as user_role, m.statut as user_membership_statut
     FROM tontines t
     LEFT JOIN memberships m ON t.id = m.tontine_id AND m.user_id = $2
     WHERE t.id = $1`,
        [id, userId]
    );

    if (rows.length === 0) return null;

    const tontine = rows[0];

    // Add stats (example of DRY view)
    const stats = await pool.query(
        `SELECT 
      (SELECT COUNT(*) FROM memberships WHERE tontine_id = $1) as total_membres,
      (SELECT SUM(montant_collecte) FROM cycles WHERE tontine_id = $1) as total_collecte`,
        [id]
    );

    tontine.stats = stats.rows[0];
    return tontine;
};

const update = async (id, tontineData) => {
    const fields = [];
    const values = [];
    let idx = 1;

    const allowedFields = ['nom', 'description', 'montant_cotisation', 'frequence', 'nombre_membres_max', 'date_debut', 'statut', 'penalites_activees', 'montant_penalite', 'code_invitation'];

    for (const [key, value] of Object.entries(tontineData)) {
        if (allowedFields.includes(key)) {
            fields.push(`${key} = $${idx}`);
            values.push(value);
            idx++;
        }
    }

    if (fields.length === 0) return null;

    values.push(id);
    const { rows } = await pool.query(
        `UPDATE tontines SET ${fields.join(', ')}, date_modification = NOW() WHERE id = $${idx} RETURNING *`,
        values
    );
    return rows[0];
};

module.exports = {
    create,
    findAll,
    findById,
    update,
};

File: mobile/lib/core/models/tontine_model.dart
class TontineModel {
  final String id;
  final String nom;
  final String? description;
  final double montantCotisation;
  final String frequence;
  final int nombreMembresMax;
  final int nombreMembresActuels;
  final String statut;
  final String? codeInvitation;
  final DateTime? dateDebut;
  final bool penalitesActivees;
  final double montantPenalite;
  final String organisateurId;
  final String? currentCycleId;

  TontineModel({
    required this.id,
    required this.nom,
    this.description,
    required this.montantCotisation,
    required this.frequence,
    required this.nombreMembresMax,
    this.nombreMembresActuels = 0,
    required this.statut,
    this.codeInvitation,
    this.dateDebut,
    this.penalitesActivees = false,
    this.montantPenalite = 0.0,
    required this.organisateurId,
    this.currentCycleId,
  });

  factory TontineModel.fromJson(Map<String, dynamic> json) {
    return TontineModel(
      id: json['id'] ?? '',
      nom: json['nom'] ?? '',
      description: json['description'],
      montantCotisation: double.tryParse(json['montant_cotisation']?.toString() ?? '0') ?? 0.0,
      frequence: json['frequence'] ?? '',
      nombreMembresMax: json['nombre_membres_max'] ?? 0,
      nombreMembresActuels: json['nombre_membres_actuels'] ?? 0,
      statut: json['statut'] ?? 'en_attente',
      codeInvitation: json['code_invitation'],
      dateDebut: json['date_debut'] != null ? DateTime.tryParse(json['date_debut'].toString()) : null,
      penalitesActivees: json['penalites_activees'] ?? false,
      montantPenalite: double.tryParse(json['montant_penalite']?.toString() ?? '0') ?? 0.0,
      organisateurId: json['organisateur_id'] ?? '',
      currentCycleId: json['current_cycle_id'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'nom': nom,
      'description': description,
      'montant_cotisation': montantCotisation,
      'frequence': frequence,
      'nombre_membres_max': nombreMembresMax,
      'date_debut': dateDebut?.toIso8601String(),
      'penalites_activees': penalitesActivees,
      'montant_penalite': montantPenalite,
    };
  }
}
